ext/nio4r/org/nio4r/ByteBuffer.java

```diff
@@ -0,0 +1,283 @@
+
+/*
+created by Upekshej
+ */
+public class ByteBuffer extends RubyObject {
+
+    private java.nio.ByteBuffer byteBuffer;
+    private String currentWritePath = "";
+    private String currentReadPath = "";
+
+    private FileChannel currentWriteFileChannel;
+    private FileOutputStream fileOutputStream;
+
+    private FileInputStream currentReadChannel;
+    private FileChannel inChannel;
+
+    public ByteBuffer(final Ruby ruby, RubyClass rubyClass) {
+        super(ruby, rubyClass);
+    }
+
+    @JRubyMethod
+    public IRubyObject initialize(ThreadContext context, IRubyObject value, IRubyObject offset, IRubyObject length) {
+        Ruby ruby = context.getRuntime();
+        if (value == ruby.getNil())
+            throw new IllegalArgumentException();
+
+        if (value instanceof RubyString) {
+            if (offset != ruby.getNil() && length != ruby.getNil()) {
+                int arrayOffset = RubyNumeric.num2int(offset);
+                int arrayLimit = RubyNumeric.num2int(length);
+                byteBuffer = java.nio.ByteBuffer.wrap(value.asJavaString().getBytes(), arrayOffset, arrayLimit);
+            } else
+                byteBuffer = java.nio.ByteBuffer.wrap(value.asJavaString().getBytes());
+        } else if (value instanceof RubyInteger) {
+            int allocationSize = RubyNumeric.num2int(value);
+            byteBuffer = java.nio.ByteBuffer.allocate(allocationSize);
+        } else {
+            throw new IllegalArgumentException();
+        }
+        return this;
+    }
+
+    /**
+     * Currently assuming only strings will come..
+     *
+     * @param context
+     * @return
+     */
+    @JRubyMethod(name = "<<")
+    public IRubyObject put(ThreadContext context, IRubyObject str) {
+        String string = str.asJavaString();
+
+        if (byteBuffer == null)
+            byteBuffer = java.nio.ByteBuffer.wrap(string.getBytes());
+        byteBuffer.put(string.getBytes());
+        return this;
+    }
+
+    //https://www.ruby-forum.com/topic/3731325
+    @JRubyMethod(name = "get")
+    public IRubyObject get(ThreadContext context) {
+        ArrayList<Byte> temp = new ArrayList<Byte>();
+        while (byteBuffer.hasRemaining()) {
+            temp.add(byteBuffer.get());
+        }
+        // String returnString = new String(toPrimitives(temp));
+
+        return JavaUtil.convertJavaToRuby(context.getRuntime(), new String(toPrimitives(temp)));
+    }
+
+    @JRubyMethod(name = "read_next")
+    public IRubyObject readNext(ThreadContext context, IRubyObject count) {
+        int c = RubyNumeric.num2int(count);
+        if (c < 1)
+            throw new IllegalArgumentException();
+        if (c <= byteBuffer.remaining()) {
+            ArrayList<Byte> temp = new ArrayList<Byte>();
+
+            while (c > 0) {
+                temp.add(byteBuffer.get());
+                c = c - 1;
+            }
+            return JavaUtil.convertJavaToRuby(context.getRuntime(), new String(toPrimitives(temp)));
```

Comment:
to avoid the `Byte[] -> char[] -> byte[]` conversions do `context.runtime.newString((ByteList) temp)`